### PR TITLE
Banek

### PR DESCRIPTION
**Pull request by dbingham**
Just creating this so GitHub will let me add comments, etc.

### CODE REVIEW

File: src/main/scala/com/oag/interview/CsvTransformer.scala [hunk 1]
```diff
@@ -61,156 +63,85 @@ import scala.util.Try
...
-  private def cum(prodHeaderIndex: Int)(slice: Seq[Array[String]]): Double = {
-    slice.foldLeft(0.0) { case (cumProd, daily) =>
-      cumProd + safeToDouble(daily(prodHeaderIndex))
+  // INTERFACE
+  def dailyToOperatedDay(prodTypes: Seq[String],
+                         outputFile: Path,
+                         prodIntervals: Seq[Int] = Seq(180, 360, 720)): Unit = {
+    managed(new PrintWriter(Files.newBufferedWriter(outputFile))) apply { writer =>
+      writer.println((wellIdColumn +: (for (pt <- prodTypes; pi <- prodIntervals) yield s"cum_${pt}_$pi")).mkString(","))
+      val prodSummary = summarize(prodTypes, prodIntervals)
```
Comment (dbingham) [inline src/main/scala/com/oag/interview/CsvTransformer.scala]: Another consideration here is the memory requirement.  Your comment above about calculating the _oldest_ intervals instead of the _newest_ is a good one and has some implications, I think, regarding approach.  For example, instead of keeping everything in memory for all wells, all the way up until the end, I think you could probably come up with a way to output "finished" information, drop it from memory, ignore irrelevant rows, and only pick back up "accumulating" information when you need to start in on the next calculation.  Obviously, depending on assumptions you make about underlying data, this could have greater or lesser impact on overall memory requirement.  For example, with that approach, the worst case is a perfectly interleaved file where you have to keep all wells in memory at once while you accumulate up to the first interval.  But, best case could be significantly better.

File: src/test/scala/com/oag/interview/CsvTest.scala [hunk 2]
```diff
@@ -67,13 +74,84 @@ class CsvTest extends FlatSpec with Matchers {
...
-  // TODO: Add a test that calls your new method for Oil and Gas and verifies appropriate output
+  "production by operated day" should "be calculated correctly" in {
+    checkCalc("gas", 365)
+    checkCalc("oil", 42)
+    checkCalc("water", 17)
+  }
 
-  // TODO: Add a test that calls your new method for Oil and Water and verifies appropriate output
+  // sampling should be sufficient given current algorithm which does not vary by interval or production type (check gas for one interval)
+  def checkCalc(prodType: String, interval: Int): Unit = {
```
Comment (dbingham) [inline src/test/scala/com/oag/interview/CsvTest.scala]: There is a bit of a chicken & egg problem here, I think.  That is, how do you ensure _this_ function calculates things correctly?  One approach might be to do a different summation implementation.  The chance of both the production function and your test function having the same bugs in the same places would be very low in that case.  But, if you _do_ see a test failure, you have to debug two things to figure out what is actually wrong.

In general, I think that using contrived data in tests works better.  That way you can contrive whatever input data you need in order to verify what you are trying to verify.  For example, you could easily contrive input data that proves boundary conditions around 0 producing days, negative intervals, incorrect production types, etc, etc.  And you can test both positive and negative conditions that way easily.

File: src/main/scala/com/oag/interview/CsvTransformer.scala [hunk 3]
```diff
@@ -61,156 +63,85 @@ import scala.util.Try
...
-      lastInterval += cum(prodIndex)(rows)
-      s"${(math rint (lastInterval * 100)) / 100}"
-    }
-  }
 
-  private def cum(prodHeaderIndex: Int)(slice: Seq[Array[String]]): Double = {
-    slice.foldLeft(0.0) { case (cumProd, daily) =>
-      cumProd + safeToDouble(daily(prodHeaderIndex))
+  // INTERFACE
+  def dailyToOperatedDay(prodTypes: Seq[String],
```
Comment (dbingham) [inline src/main/scala/com/oag/interview/CsvTransformer.scala]: I was a bit evil in the setup on this point.  There is a philosophical debate around this.  I made the datatype for the production type a `String` in the starting point code specifically so I could ask this question and see what people say.  There are good arguments on both sides, so I won't spoil it with my opinion, but I will spring the trap I set and would be interested to hear your thoughts...

Since the input format is known and all of the parsing code is encapsulated here, why use a `String` as the input type?  Doesn't a pure string matching approach open you up to a class of bugs in callers where they pass in Strings that don't match expected inputs?  What happens in your implementation if a caller asks for a summary of `Seq("oli", "gas", "watar")`?  What _should_ it do in that case?

File: src/test/scala/com/oag/interview/CsvTest.scala [hunk 4]
```diff
@@ -67,13 +74,84 @@ class CsvTest extends FlatSpec with Matchers {
...
+
+  // test that calls your new method for Oil, Gas, and Water and verifies appropriate output
+  "oilgaswater-by-operated-day" should "contain appropriate columns" in {
+    // go ahead and allow the helper function to test the default parameter of the CsvTransformer interface
+    // not the best way to do this, but in the interests of making checkColumns more generic, allowing this
+    // - if we screw up and pass a non-default set of intervals, the test will fail anyway, so no big deal
+    checkColumns(Seq("oil", "gas", "water"), Seq(180,360,720), defaultIntervals = true)
+  }
+
+  def checkColumns(prodTypes: Seq[String], intervals: Seq[Int], defaultIntervals: Boolean=false): Unit = {
```
Comment (dbingham) [inline src/test/scala/com/oag/interview/CsvTest.scala]: Same here as above in `checkCalc`

**Comment by dbingham** [comment]
Dang, sorry the order is all over the place on that.  I assumed it would attach the comments in code order, not the order I typed them.

File: README.md [hunk 5]
```diff
@@ -74,7 +74,7 @@ Build and test with this command: `./sbtd clean test`.  For details, see `docker
   appropriate.  Imagine that this is part of a larger system with many similar types of processing happening in various
   contexts.
 
-* Keep [SOLID Design Principles](https://en.wikipedia.org/wiki/SOLID_(object-oriented_design)) in mind while working. 
+* Keep [SOLID Design Principles](https://en.wikipedia.org/wiki/SOLID_(object-oriented_design\)) in mind while working.
```
Comment (dbingham) [inline README.md]: Interestingly, this change doesn't work for me.  In the upstream repo, browsing through GitHub, the link works.  In your branch here in your repo, GitHub doesn't render the link correctly.

Comment (natonic77) [reply README.md]: You know, I wonder if this is a bug (or at least difference in interpretation) in the IntelliJ markdown renderer.  That is the only place I tested the README.md file and that is where the link didn't work.  I do see that my change doesn't work in github.  Weird.

File: src/main/scala/com/oag/interview/CsvTransformer.scala [hunk 6]
```diff
@@ -61,156 +63,85 @@ import scala.util.Try
...
-      val production = Source.fromInputStream(is).getLines().toSeq
-      val headers = parser.parseLine(production.head)
 
-      // confirm date parsing on 10 samples
-      val samples = production.tail.take(10).map(parser.parseLine)
-      val dateIndex = this.dateIndex(headers)
-      samples.foreach(a => parseDate(a(dateIndex)))
+  // TYPES
+  class Resource(val name: String, val col: Int, intervals: Seq[Int], val prodCount: Int, val cumByInterval: Map[Int, Long]) {
+    def this(name: String, col: Int, intervals: Seq[Int]) = this(name, col, intervals, 0, intervals.map((_, 0L)).toMap)
```
[the remark this answers is not in the document]
Comment (natonic77) [reply src/main/scala/com/oag/interview/CsvTransformer.scala]: Yes, I agree.  Took the short path, but it is a less robust approach than separating the concerns.  I'm going to take a look at the alternate constructor syntax some more.  I'd like to understand that approach.

File: src/main/scala/com/oag/interview/CsvTransformer.scala [hunk 7]
```diff
@@ -61,156 +63,85 @@ import scala.util.Try
...
 
-      headers
+    def accumulate(parsedProd: Array[String]): Resource = {
+      val prod = safeToDouble(parsedProd(col)).round
+      val updatedCumByInterval =
+        (for (i <- cumByInterval.keys) yield
+          (i, if (prod > 0 && prodCount < i) cumByInterval(i) + prod else cumByInterval(i))).toMap
+      if (prod > 0L) new Resource(name, col, intervals, prodCount + 1, updatedCumByInterval) else this
```
[the remark this answers is not in the document]
Comment (natonic77) [reply src/main/scala/com/oag/interview/CsvTransformer.scala]: gp

File: src/main/scala/com/oag/interview/CsvTransformer.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (natonic77) [reply src/main/scala/com/oag/interview/CsvTransformer.scala]: Yes, good point on the structure.  Separation of concerns on meta and state would be more robust.  Also going to take a look at the alternate constructor syntax as I'd like to understand that more.

File: src/main/scala/com/oag/interview/CsvTransformer.scala [hunk 8]
```diff
@@ -61,156 +63,85 @@ import scala.util.Try
...
+                        wellIdx: Int, resources: Map[String, Resource], intervals: Seq[Int], maxInterval: Int): ProductionAccumulator = {
+      if (!lines.hasNext) acc
+      else {
+        val parsed = parser.parseLine(lines.next)
+        if (parsed(wellIdx).isEmpty) summarizeHelper(lines, acc, wellIdx, resources, intervals, maxInterval)
+        else summarizeHelper(lines,
+          acc + (parsed(wellIdx) -> (for (rsrc <- acc.getOrElse(parsed(wellIdx), resources)) yield (rsrc._1, rsrc._2.accumulate(parsed)))),
+          wellIdx, resources, intervals, maxInterval)
       }
     }
```
[the remark this answers is not in the document]
Comment (natonic77) [reply src/main/scala/com/oag/interview/CsvTransformer.scala]: Definitely was following an Odersky pattern here, but I'm going to take a look at other ways to do it from a learning standpoint.  Are you suggesting maybe a more traditional loop with mutating local state would be appropriate here?  As I understand it many of the collection manipulation functions in scala use recursion, so using one of those might not avoid the problem.

File: src/main/scala/com/oag/interview/CsvTransformer.scala [hunk 9]
```diff
@@ -61,156 +63,85 @@ import scala.util.Try
...
-    }
-  }
 
-  // production per wellId
-  private def prodPerWellId(wellId: String): Seq[Array[String]] = {
-    managed(Files.newInputStream(inputFile)) apply { is =>
-      Source.fromInputStream(is).getLines().filter(_.contains(wellId)).map(parser.parseLine).toList
-    }
+    val lines = Source.fromInputStream(Files.newInputStream(inputFile)).getLines()
+    val parsedHeader = parser.parseLine(lines.next)
```
[the remark this answers is not in the document]
Comment (natonic77) [reply src/main/scala/com/oag/interview/CsvTransformer.scala]: I'm sure you are correct, and that points to a different unit test I could have created.  This reminds me, I meant to ask about the resource.manage object.  I am not familiar with that namespace or object.  I tried looking it up and couldn't find any documentation anywhere.  Can you point me to something on that?

File: src/main/scala/com/oag/interview/CsvTransformer.scala [hunk 10]
```diff
@@ -61,156 +63,85 @@ import scala.util.Try
...
-
-          writer.println((wellId +: cumProdPerInterval).mkString(","))
+  type ProductionAccumulator = Map[String, Map[String, Resource]]
+
+
+  // PROCESSING
+  private def summarize(requested: Seq[String], intervals: Seq[Int]): ProductionAccumulator = {
+    @tailrec
+    def summarizeHelper(lines: Iterator[String], acc: ProductionAccumulator,
+                        wellIdx: Int, resources: Map[String, Resource], intervals: Seq[Int], maxInterval: Int): ProductionAccumulator = {
```
[the remark this answers is not in the document]
Comment (natonic77) [reply src/main/scala/com/oag/interview/CsvTransformer.scala]: You are correct, I had added it for a future optimization (interrupting the file traversal once all intervals were achieved), but decided not to do the optimization because the benefit seemed likely to be low in comparison to the extra complexity of implementing and maintaining it.

File: src/main/scala/com/oag/interview/CsvTransformer.scala (hunk 1 above)
Comment (natonic77) [reply src/main/scala/com/oag/interview/CsvTransformer.scala]: Yes, in fact in an earlier iteration when I misunderstood the problem, I was doing something akin to what you describe here.  When I re-read the problem and implemented this solution, I didn't go back and optimize for this.  It's a good point that it would at least make the algorithm more scalable for massive well counts (or a massive combination of wells and intervals).

File: src/test/scala/com/oag/interview/CsvTest.scala (hunk 2 above)
Comment (natonic77) [reply src/test/scala/com/oag/interview/CsvTest.scala]: Yeah, I was worried about the complexity of this calculation test, and guessed that was at least part of what it wasn't in the original code.  I should have thought a little deeper about the test environment.  I was really wanting to modify the source data for the tests, but CsvCreator was off limits.  If I'd thought a little harder though I would have realized there was no reason I couldn't do exactly as you suggest and contrive additional input data to drive the calculations more simply (and as you point out limit the space from which potential errors would come from).

File: src/main/scala/com/oag/interview/CsvTransformer.scala (hunk 3 above)
Comment (natonic77) [reply src/main/scala/com/oag/interview/CsvTransformer.scala]: To be honest, this did cross my mind, and I almost added a test to ask for pecans.  In answer to the question about what happens in my implementation with unexpected inputs, I expect it would run into trouble indexing the CSV header.  I didn't end up doing the pecan test or making the changes to be more robust about unexpected headers, but if I were to stick with string inputs, I would add validation on the inputs.  I think it would be acceptable in some contexts to fail immediately if the validation failed (if expecting to wait a long time for calculation to complete, fail fast is good), but in other contexts, I imagine it would be preferable to ignore unexpected inputs and go ahead and do the full processing on valid inputs (maybe partial information now is more important than all the information later after fixing messed up inputs).

I ended up taking a bit of an easy out on it by accepting the assumptions of the original interface (which I recognize I didn't have to do), but if I were exposing this call as an API to a third party, I would go with the symbolic approach over the string approach in this context.  The production types we have to deal with in the O&G context are a pretty constrained set, so registering and maintaining new ones shouldn't be that burdensome, and it would make life a lot easier for the API consumer.  That said, there is still the problem of corrupted source data (what if the caller asked for GAS, but the source data had "gsa"?), but handling this would be necessary regardless of using the String or symbolic approach.

This discussion leads to another question I had and did not test, what if the requested interval size is larger than the file?  One option would be to take what we had and some how mark the output (maybe add an asterisk to that summary cell) or to report that cell as N/A so as not to mislead the consumer of the summary.

File: src/test/scala/com/oag/interview/CsvTest.scala (hunk 4 above)
Comment (natonic77) [reply src/test/scala/com/oag/interview/CsvTest.scala]: I was a lot less worried about this one than the other (see my comment on lines 139 to 141 for what was going through my mind in regards to one error-prone aspect of this function - alignment of the intervals passed in and use of defaultIntervals=true).

Other than the default interval test weirdness addressed in my comment (which I'm not sure I'd leave in there at all if this were a public API), there is very little processing code here that I would be very worried about being the source of a hard to find error in the test code.  Unlike the checkCalc function above, I'd probably leave this one largely as it is in a production system.  Interested if you have other thoughts, though.

File: src/main/scala/com/oag/interview/CsvTransformer.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (natonic77) [reply src/main/scala/com/oag/interview/CsvTransformer.scala]: Looked at Alvin Alexander's article on tail recursion and at Oderskey's discussion of it, they both seem pretty confident that if the compiler thinks your function is tail recursive (as proved with @tailrec annotation) it will execute the recursion in a single stack frame.  Would be interesting to find out if there are exceptions to this, but the argument seems to make sense - and Alexander provides an additional way to prove at the end of his page: https://alvinalexander.com/scala/fp-book/tail-recursive-algorithms.

File: src/main/scala/com/oag/interview/CsvTransformer.scala [hunk 11]
```diff
@@ -61,156 +63,85 @@ import scala.util.Try
...
+  // TYPES
+  class Resource(val name: String, val col: Int, intervals: Seq[Int], val prodCount: Int, val cumByInterval: Map[Int, Long]) {
+    def this(name: String, col: Int, intervals: Seq[Int]) = this(name, col, intervals, 0, intervals.map((_, 0L)).toMap)
 
-      headers
+    def accumulate(parsedProd: Array[String]): Resource = {
+      val prod = safeToDouble(parsedProd(col)).round
+      val updatedCumByInterval =
+        (for (i <- cumByInterval.keys) yield
+          (i, if (prod > 0 && prodCount < i) cumByInterval(i) + prod else cumByInterval(i))).toMap
```
[the remark this answers is not in the document]
Comment (natonic77) [reply src/main/scala/com/oag/interview/CsvTransformer.scala]: Apparently this conversion is exactly what the compiler does: https://stackoverflow.com/questions/14598990/confused-with-the-for-comprehension-to-flatmap-map-transformation.

File: src/main/scala/com/oag/interview/CsvTransformer.scala [hunk 12]
```diff
@@ -61,156 +63,85 @@ import scala.util.Try
...
 
-  // production per wellId
-  private def prodPerWellId(wellId: String): Seq[Array[String]] = {
-    managed(Files.newInputStream(inputFile)) apply { is =>
-      Source.fromInputStream(is).getLines().filter(_.contains(wellId)).map(parser.parseLine).toList
-    }
+    val lines = Source.fromInputStream(Files.newInputStream(inputFile)).getLines()
+    val parsedHeader = parser.parseLine(lines.next)
+    summarizeHelper(lines,
+      Map[String, Map[String, Resource]](),
```
[the remark this answers is not in the document]
Comment (natonic77) [reply src/main/scala/com/oag/interview/CsvTransformer.scala]: You are correct.  I think I tried this somewhere else and IntelliJ didn't like it, so I stopped trying.  Glad to see it does work though, as I didn't like duplicating the type signature.
